### PR TITLE
[MXNET-416] Add docker cache

### PR DESCRIPTION
**Pull request by marcoabreu**
## Description ##
This PR introduces shared Docker cache to our Docker pipeline. A Jenkins job generates all the Docker cache upon every new commit to master and publishes it to an S3 bucket. All slaves retrieve this cache if necessary instead of rebuilding the Docker layers.
This cache is part of the auto scaling project. It is necessary to reduce the cold start time of our slaves. At the moment, most time is spent by generating the Docker layers instead of the actual execution. The retrieval of the cache takes less than a minute if run inside EC2, making this an efficient approach.

All required steps for setup are described in https://cwiki.apache.org/confluence/display/MXNET/Setup#Setup-Dockercache. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Add Jenkinsfile which will be run every time a new commit is being pushed to master
- Reorganized main Jenkinsfile to have one function to call build.py and pass the cache parameters
- Reorganized Dockerfiles to improve caching efficiency due to uncacheable build args.

## Comments ##
- Rebuild of all Dockerfiles (in parallel) takes 30 minutes max. Since most of the time is spent on IO, there's no harm to parallelize the entire process on a software level without having to distribute the job across multiple slaves.
- Cache load time due to cache miss is between 30 and 120 seconds (10 sec download, rest is load overhead)
- Cache hit introduces < 1 second overhead
- This PR can only be merged after the new set of slaves is deployed as part of the auto scaling project

### CODE REVIEW

File: ci/Jenkinsfile_docker_cache [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Jenkins pipeline to generate the centralized docker cache
+// See documents at https://jenkins.io/doc/book/pipeline/jenkinsfile/
+
+// timeout in minutes
+max_time = 120
```
Comment (marcoabreu) [inline ci/Jenkinsfile_docker_cache]: Add timeout

File: ci/Jenkinsfile_docker_cache [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+try {
+  stage("Docker cache build & publish") {
+    node('mxnetlinux-cpu') {
+      ws('workspace/docker_cache') {
+        init_git()
+        sh "pip3 install --user virtualenv"
+        sh "python3 -m virtualenv virtualenv_docker_cache"
+        sh ". virtualenv_docker_cache/bin/activate"
+        sh "ls ci"
```
Comment (marcoabreu) [inline ci/Jenkinsfile_docker_cache]: remove

File: ci/Jenkinsfile_docker_cache [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  }
+}
+
+
+try {
+  stage("Docker cache build & publish") {
+    node('mxnetlinux-cpu') {
+      ws('workspace/docker_cache') {
+        init_git()
+        sh "pip3 install --user virtualenv"
```
Comment (marcoabreu) [inline ci/Jenkinsfile_docker_cache]: Consilidate into one script

Comment (marcoabreu) [reply ci/Jenkinsfile_docker_cache]: Not possible due to virtualenv..

File: ci/Jenkinsfile_docker_cache [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+
+try {
+  stage("Docker cache build & publish") {
+    node('mxnetlinux-cpu') {
+      ws('workspace/docker_cache') {
+        timeout(time: max_time, unit: 'MINUTES') {
+          init_git()
+          //sh "pip3 install --user virtualenv"
```
Comment (eric-haibin-lin) [inline ci/Jenkinsfile_docker_cache]: Remove unused code?

File: ci/build.py [hunk 5]
```diff
@@ -33,6 +33,7 @@
 import shutil
 import subprocess
 import sys
+import docker_cache
```
Comment (eric-haibin-lin) [inline ci/build.py]: Some users are asking if mxnet maintains any official docker containers. This is probably only used for CI, but i guess some of these procedures can be reused if we want to host official docker containers?

File: ci/docker_cache.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    if len(platforms) == 0:
+        return
+
+    Parallel(n_jobs=len(platforms), backend="multiprocessing")(
+        delayed(_build_save_container)(platform, bucket)
+        for platform in platforms)
+
+    is_error = False
+
+    logging.error('TODO: implement error detection')
```
Comment (eric-haibin-lin) [inline ci/docker_cache.py]: Is this going to be included in this PR?

File: ci/Jenkinsfile_docker_cache (hunk 4 above)
Comment (marcoabreu) [reply ci/Jenkinsfile_docker_cache]: Done

File: ci/docker_cache.py (hunk 6 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: Jenkinsfile [hunk 7]
```diff
@@ -92,24 +92,34 @@ echo ${libs} | sed -e 's/,/ /g' | xargs md5sum
 """
 }
 
+def docker_run(platform, function_name, use_nvidia, shared_mem = '500m') {
+  def command = "ci/build.py --download-docker-cache --docker-cache-bucket ${env.DOCKER_CACHE_BUCKET} %USE_NVIDIA% --platform %PLATFORM% --shm-size %SHARED_MEM% /work/runtime_functions.sh %FUNCTION_NAME%"
+  command = command.replaceAll('%USE_NVIDIA%', use_nvidia ? '--nvidiadocker' : '')
+  command = command.replaceAll('%PLATFORM%', platform)
+  command = command.replaceAll('%FUNCTION_NAME%', function_name)
+  command = command.replaceAll('%SHARED_MEM%', shared_mem)
```
Comment (KellenSunderland) [inline Jenkinsfile]: Can you explain a bit why the shared mem is needed here?

File: Jenkinsfile [hunk 8]
```diff
@@ -92,24 +92,34 @@ echo ${libs} | sed -e 's/,/ /g' | xargs md5sum
...
+
+  sh command
+}
+
 // Python unittest for CPU
 // Python 2
 def python2_ut(docker_container_name) {
   timeout(time: max_time, unit: 'MINUTES') {
-    sh "ci/build.py --platform ${docker_container_name} /work/runtime_functions.sh unittest_ubuntu_python2_cpu"
+    docker_run(docker_container_name, 'unittest_ubuntu_python2_cpu', false)
```
Comment (KellenSunderland) [inline Jenkinsfile]: While this is nice from a readability point of view, it goes backwards on our efforts to have easily reproducible commands locally.

File: ci/Jenkinsfile_docker_cache [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Jenkins pipeline to generate the centralized docker cache
+// See documents at https://jenkins.io/doc/book/pipeline/jenkinsfile/
+
+// timeout in minutes
+max_time = 120
```
Comment (larroy) [inline ci/Jenkinsfile_docker_cache]: can we clarify which timeout is this? preferrably in variable name.

File: ci/Jenkinsfile_docker_cache [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+err = null
+
+// initialize source codes
+def init_git() {
+  deleteDir()
+  retry(5) {
+    try {
+      // Make sure wait long enough for api.github.com request quota. Important: Don't increase the amount of
+      // retries as this will increase the amount of requests and worsen the throttling
+      timeout(time: 15, unit: 'MINUTES') {
```
Comment (larroy) [inline ci/Jenkinsfile_docker_cache]: could we put it in the top of the file, as it's a constant if I understand correctly?

File: Jenkinsfile (hunk 7 above)
Comment (marcoabreu) [reply Jenkinsfile]: Shared memory was introduced by @ThomasDelteil. It is required for the tutorial tests

File: Jenkinsfile (hunk 8 above)
Comment (marcoabreu) [reply Jenkinsfile]: The commands are being printed out in blue ocean, making them still copy&pasteable. The problem here is that otherwise we'd have too much copy&paste. What would you propose?

File: ci/Jenkinsfile_docker_cache (hunk 9 above)
Comment (marcoabreu) [reply ci/Jenkinsfile_docker_cache]: Done

File: ci/Jenkinsfile_docker_cache (hunk 10 above)
Comment (marcoabreu) [reply ci/Jenkinsfile_docker_cache]: Done

File: ci/docker_cache.py [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+    """
+    global cached_aws_session
+    if cached_aws_session:
+        return cached_aws_session
+
+    session = boto3.Session()  # Uses IAM user credentials
+    cached_aws_session = session
+    return session
+
+def _format_docker_cache_filepath(output_dir, docker_tag):
```
Comment (larroy) [inline ci/docker_cache.py]: missing return type here for consistency

File: ci/docker_cache.py [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+                logging.info('Docker cache for {} loaded successfully'.format(docker_tag))
+            except CalledProcessError as e:
+                logging.error('Error during load of docker cache for {} at {}'.format(docker_tag, tar_file_path))
+                logging.exception(e)
+                return
+    else:
+        logging.info('No cached remote image of {} present'.format(docker_tag))
+
+
+def _docker_layer_exists(layer_id):
```
Comment (larroy) [inline ci/docker_cache.py]: missing return type annotation

File: ci/docker_cache.py [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == "404":
+            logging.debug('{} does not exist in S3 yet'.format(s3_object.key))
+        else:
+            raise
+
+    return None
+
+
+def load_docker_cache(bucket_name, docker_tag):
```
Comment (larroy) [inline ci/docker_cache.py]: missing return type

Comment (larroy) [reply ci/docker_cache.py]: missing documentation

File: ci/docker_cache.py [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+        _compile_upload_cache_file(bucket_name=bucket, docker_tag=docker_tag, image_id=image_id)
+        return None
+    except Exception:
+        logging.exception('Unexpected exception during build of {}'.format(docker_tag))
+        return platform
+        # Error handling is done by returning the errorous platform name. This is necessary due to
+        # Parallel being unable to handle exceptions
+
+
+def _compile_upload_cache_file(bucket_name, docker_tag, image_id):
```
Comment (larroy) [inline ci/docker_cache.py]: return type and doc

File: ci/docker_cache.py [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+    is_error = False
+    for platform_result in platform_results:
+        if platform_result is not None:
+            logging.error('Failed to generate {}'.format(platform_result))
+            is_error = True
+
+    return 1 if is_error else 0
+
+
+def _build_save_container(platform, bucket):
```
Comment (larroy) [inline ci/docker_cache.py]: missing return type

File: ci/build.py [hunk 16]
```diff
@@ -61,17 +62,44 @@ def get_docker_binary(use_nvidia_docker: bool) -> str:
...
+    :param platform: Platform
+    :param docker_binary: docker binary to use (docker/nvidia-docker)
+    :return: Id of the top level image
+    """
+
     tag = get_docker_tag(platform)
     logging.info("Building container tagged '%s' with %s", tag, docker_binary)
     cmd = [docker_binary, "build",
         "-f", get_dockerfile(platform),
+        "--rm=false",
```
[the remark this answers is not in the document]
Comment (marcoabreu) [reply ci/build.py]: We won't to keep the intermediary layers to prime the build cache. Otherwise, we'll end up in missing layers if you execute ``docker history IMAGE_NAME``.

File: ci/docker_cache.py (hunk 11 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py (hunk 12 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py (hunk 13 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py (hunk 14 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py (hunk 15 above)
Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+cached_aws_session = None
+
+
+class ProgressPercentage(object):
+    def __init__(self, object_name, size):
+        self._object_name = object_name
+        self._size = size
+        self._seen_so_far = 0
+        self._last_percentage = 0
+        self._lock = threading.Lock()
```
Comment (larroy) [inline ci/docker_cache.py]: why do we need this lock? where is realeased?

File: ci/docker_cache.py [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+Utility to handle distributed docker cache
```
Comment (larroy) [inline ci/docker_cache.py]: could we put one or two phrases describing the approach?

Comment (marcoabreu) [reply ci/docker_cache.py]: Done

File: ci/docker_cache.py [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+            s3_object.download_file(
+                Filename=tar_file_path,
+                Callback=ProgressPercentage(object_name=docker_tag, size=s3_object.content_length))
+
+            # Load layers
+            cmd = ['docker', 'load', '-i', tar_file_path]
+            try:
+                check_call(cmd)
+                logging.info('Docker cache for {} loaded successfully'.format(docker_tag))
+            except CalledProcessError as e:
```
[the remark this answers is not in the document]
Comment (marcoabreu) [reply ci/docker_cache.py]: Exactly

File: Jenkinsfile (hunk 8 above)
Comment (marcoabreu) [reply Jenkinsfile]: Is your focus to have the commands reproducible when looking at the Jenkinsfile or when looking at a past builds? The commands are copy&pasteable when looking at Blue Ocean? I don't feel strongly about this, so I don't mind changing this back.

File: ci/Jenkinsfile_docker_cache [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+def init_git() {
+  deleteDir()
+  retry(5) {
+    try {
+      // Make sure wait long enough for api.github.com request quota. Important: Don't increase the amount of
+      // retries as this will increase the amount of requests and worsen the throttling
+      timeout(time: git_timeout, unit: 'MINUTES') {
+        checkout scm
+        sh 'git submodule update --init --recursive'
+        sh 'git clean -d -f'
```
Comment (lebeg) [inline ci/Jenkinsfile_docker_cache]: To have a really clean build you should use the -x option as well to remove otherwise ignored build artifacts

File: ci/build.py [hunk 21]
```diff
@@ -61,17 +62,44 @@ def get_docker_binary(use_nvidia_docker: bool) -> str:
...
+    # check_call would have failed
+    image_id = _get_local_image_id(docker_binary=docker_binary, docker_tag=tag)
+    if not image_id:
+        raise FileNotFoundError('Unable to find docker image id matching with {}'.format(tag))
+    return image_id
+
+
+def _get_local_image_id(docker_binary, docker_tag):
+    """
+    Get the image if of the local docker layer with the passed tag
```
Comment (lebeg) [inline ci/build.py]: Can you rephrase? What is the 'if' for?

Comment (marcoabreu) [reply ci/build.py]: Sorry, was a typo

File: ci/Jenkinsfile_docker_cache (hunk 20 above)
Comment (marcoabreu) [reply ci/Jenkinsfile_docker_cache]: Changed
